src/main/java/org/truffleruby/core/array/ArraySliceNode.java

```diff
@@ -41,9 +41,7 @@ public DynamicObject readInBounds(DynamicObject array,
         if (emptyArray.profile(from >= to)) {
             return createArray(null, 0);
         } else {
-            final ArrayMirror arrayMirror = strategy.newMirror(array);
-            final Object slice = arrayMirror.extractRange(from, to).getArray();
-            strategy.makeStorageImmutable(array);
+            final Object slice = strategy.makeStorageShared(array).extractRange(from, to).getArray();
```

Comment:
Is this correct/enough? What if the original array is modified?

Reply:
Yes this is correct `makeSotrageShared` makes the original array's storage `DelegatedArrayStorage`(or will when we switch that feature). Any future attempt to modify the slice or the original will cause the contents to be CoW'd and the other copy won't see the changes. There's a spec that checks for modification of slices, so if we get it wrong it gets caught pretty quickly.

Reply:
Ah, nice, so the part I was missing is currently makeStorageShared is a no-op but will change to mark the storage of `array` shared as well (a DelegatedArrayStorage).
It has the side effects of adding writes on operation that would need reads, but I don't think that's problematic currently (might be complex with concurrent strategies, we'll see).